src/components/HelpScoutBeacon/useBeaconSuggestions.js

```diff
@@ -0,0 +1,58 @@
+import { useEffect, useCallback, useState } from 'react'
+import suggestions from './suggestions'
+import { staticApps } from '../../static-apps'
+
+const sectionToSuggestions = new Map(suggestions)
+
+function useBeaconSuggestions({
+  apps,
+  beaconReady,
+  locator: { instanceId, path },
+  optedIn,
+}) {
+  const [originalOptedIn] = useState(optedIn)
+  const [canSuggest, setCanSuggest] = useState(optedIn)
+  const [shouldSuggest, setShouldSuggest] = useState(false)
+  const getSection = useCallback(() => {
+    if (path === '/') {
+      return 'onboarding'
+    }
+    if (staticApps.has(instanceId)) {
+      return instanceId
+    }
+    const app = apps.find(({ proxyAddress }) => proxyAddress === instanceId)
+    if (app) {
+      return app.name.toLowerCase()
+    }
+    return null
+  }, [path, instanceId, apps])
+
+  useEffect(() => {
+    if (!shouldSuggest) {
+      return
+    }
+    window.Beacon('suggest', sectionToSuggestions.get(getSection()))
+  }, [getSection, shouldSuggest])
+
+  useEffect(() => {
+    if (!canSuggest || !beaconReady) {
+      return
+    }
+    // this only happens when user opts in
+    // when opting in beaconReady is set after the open event has been triggered
+    // give it a minute before suggesting or a weird reace condition happens
+    let timeout
+    if (!originalOptedIn) {
+      timeout = setTimeout(() => setShouldSuggest(true), 1000)
+      return
```

Comment:
Will delegate to hook-fu master @bpierre to determine if this is a potential memory leak (since on this code path it's not returning the clean up function).

It feels like writing `useEffect()` hooks should use a single point of return.

Reply:
Yes we should return the clean up function here, otherwise it will only run when there is no timeout. :+1:

Reply:
_hook fu_ master, noice 😄 
It only happens once.

Reply:
Even if it happens once, it’s still useful to clean it up properly, e.g. if for hot reload to work properly (one day :D), or if the parent gets unmounted / remounted for any reason, and also as a good practice in general, the same way we use `componentWillUnmount()` to clean up components that never unmount.

At the moment, the cleanup function can only run with `timeout === undefined` (we might as well not have it :smile:), since we return right after assigning it.

Reply:
No, no, I mentioned it only happened once for context but I agree on adding the return function (it is already there).

Reply:
Oh ok, sorry I misunderstood haha :see_no_evil: